llvm/lib/Transforms/AggressiveInstCombine/AggressiveInstCombine.cpp

```diff
@@ -1224,6 +1224,69 @@ static bool foldLibCalls(Instruction &I, TargetTransformInfo &TTI,
   return false;
 }
 
+static bool isSaveToNarrow(unsigned opc, uint64_t num1, uint64_t num2) {
+  if (num1 > 0xffffffff || num2 > 0xffffffff) {
+    // if `num > 0xffffffff`, then `%and = and i64 %a, num` may or may not have
+    // higher 32bit set. Which cause truncate possibly lose infomation
+    return false;
+  }
+  switch (opc) {
+  // If `%and = and i64 %a, num` where num <= 0xffffffff, then `%and` must be
+  // positive.
+  // Since add and mul both increasing function on positive integer domain and
+  // `%ai <= numi`, then if `(num1 op num2) <= 0xffffffff` we have `%a1 + %a2 <=
+  // 0xffffffff`
+  case Instruction::Add:
+    return (num1 + num2) <= 0xffffffff;
+  case Instruction::Mul:
+    return (num1 * num2) <= 0xffffffff;
+    break;
+  }
+
+  return false;
+}
+
+static bool tryNarrowMathIfNoOverflow(Instruction &I,
+                                      TargetTransformInfo &TTI) {
+  unsigned opc = I.getOpcode();
+  if (opc != Instruction::Add && opc != Instruction::Mul) {
+    return false;
+  }
+  LLVMContext &ctx = I.getContext();
+  Type *I64Type = Type::getInt64Ty(ctx);
+  Type *I32Type = Type::getInt32Ty(ctx);
+
+  if (I.getType() != I64Type || !TTI.isTruncateFree(I64Type, I32Type)) {
+    return false;
+  }
+  InstructionCost CostOp64 =
+      TTI.getArithmeticInstrCost(opc, I64Type, TTI::TCK_RecipThroughput);
+  InstructionCost CostOp32 =
+      TTI.getArithmeticInstrCost(opc, I32Type, TTI::TCK_RecipThroughput);
+  InstructionCost CostZext64 = TTI.getCastInstrCost(
+      Instruction::ZExt, I64Type, I32Type, TTI.getCastContextHint(&I),
+      TTI::TCK_RecipThroughput);
+  if ((CostOp64 - CostOp32) <= CostZext64) {
+    return false;
+  }
+  uint64_t AndConst0, AndConst1;
+  Value *X;
+  if ((match(I.getOperand(0), m_And(m_Value(X), m_ConstantInt(AndConst0))) ||
+       match(I.getOperand(0), m_And(m_ConstantInt(AndConst0), m_Value(X)))) &&
+      (match(I.getOperand(1), m_And(m_Value(X), m_ConstantInt(AndConst1))) ||
+       match(I.getOperand(1), m_And(m_ConstantInt(AndConst1), m_Value(X)))) &&
+      isSaveToNarrow(opc, AndConst0, AndConst1)) {
```

Comment:
Instead of matching this specific pattern, should use computeKnownBits to get the range